craft_cli/pytest_plugin.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2022 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Different fixtures for easier testability of Craft CLI services."""
+
+import contextlib
+import os
+import pathlib
+import re
+import tempfile
+from unittest.mock import call
+
+import pytest
+
+from craft_cli import messages
+
+
+@pytest.fixture(autouse=True)
+def init_emitter(monkeypatch):
+    """Ensure emit is always clean, and initted (in test mode).
+
+    Note that the `init` is done in the current instance that all modules already
+    acquired.
+    """
+    # init with a custom log filepath so user directories are not involved here; note that
+    # we're not using pytest's standard tmp_path as Emitter would write logs there, and in
+    # effect we would be polluting that temporary directory (potentially messing with
+    # tests, that may need that empty), so we use another one.
+    temp_fd, temp_logfile = tempfile.mkstemp(prefix="emitter-logs")
+    os.close(temp_fd)
+    temp_logfile = pathlib.Path(temp_logfile)
+
+    monkeypatch.setattr(messages, "TESTMODE", True)
+    messages.emit.init(
+        messages.EmitterMode.QUIET, "test-emitter", "Hello world", log_filepath=temp_logfile
+    )
+    yield
+    # end machinery (just in case it was not ended before; note it's ok to "double end")
+    messages.emit.ended_ok()
+    temp_logfile.unlink()
+
+
+class _RegexComparingText(str):
+    """A string that compares for equality using regex.match."""
+
+    def __eq__(self, other):
+        return bool(re.match(self, other, re.DOTALL))
+
+    def __hash__(self):
+        return str.__hash__(self)
+
+
+class _RecordingEmitter:
```

Comment:
why not make this public if referenced in the documentation?

Reply:
Because it's not meant to be used directly. But if we need it public for the documentation, I'll make it public and annotate in its docstring that is not meant to be used directly...